gcc/rust/typecheck/rust-hir-type-check-expr.h

```diff
@@ -1274,17 +1286,30 @@ class TypeCheckExpr : public TypeCheckBase
 	//
 	// Something like this
 	//
-	// bool seg_is_module = mappings->lookup_module (ref);
-	// if (seg_is_module)
-	//   {
-	//     if (have_more_segments)
-	//       continue;
-	//
-	//     rust_error_at (seg.get_locus (), "expected value");
-	//     return new TyTy::ErrorType (expr.get_mappings ().get_hirid ());
-	//   }
-
-	TyTy::BaseType *lookup = nullptr;
+        auto seg_is_module = (nullptr != mappings->lookup_module (expr.get_mappings ().get_crate_num (), ref));
+
+        if (seg_is_module)
+          {
+	    // A::B::C::this_is_a_module::D::E::F
+	    //          ^^^^^^^^^^^^^^^^
+	    //          Currently handling this.
+	    if (have_more_segments)
+              {
+                (*offset)++;
+                continue;
+              }
+
+	    // In the case of :
+	    // A::B::C::this_is_a_module
+	    //          ^^^^^^^^^^^^^^^^
+            // This is an error, we are not expecting a module.
+	    rust_error_at (seg.get_locus (), "expected value");
+            return new TyTy::ErrorType (expr.get_mappings ().get_hirid
+                                        ());
+          }
+
+        TyTy::BaseType *lookup
+	  = nullptr;
 	if (!context->lookup_type (ref, &lookup))
```

Comment:
This all looks good here. You have a good comment about the nullptr vs error node. There is a nasty bug in here when we have bad name resolution, where you hit an assertion in the caller of resolve_root_node: when we have a path like:

module::foobar

and say for example foobar is not resolved or does not exist the result of this function is going to be a nullptr. I think the fix will be tricky to fix that as part of this PR so lets keep the changes to a minimum to handle modules for now. So long as there are no regressions to the testsuite